src/communication/response_broker.hpp

```diff
@@ -82,76 +105,112 @@ namespace cubcomm
   {
     get_bucket (a_rsn).register_response (a_rsn, std::move (a_payload));
   }
 
-  template <typename T_PAYLOAD>
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::register_error (response_sequence_number a_rsn, T_ERROR &&a_error)
+  {
+    get_bucket (a_rsn).register_error (a_rsn, std::move (a_error));
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::bucket (T_ERROR a_no_error)
+    : m_no_error { a_no_error }
+  {
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::~bucket ()
+  {
+    // NOTE: might not hold in the event that a peer server crashes before consuming a response;
+    // in which case the response will linger on in the queue indefinitely
+    assert (m_response_payloads.empty ());
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
   void
-  response_broker<T_PAYLOAD>::bucket::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
   {
     {
       std::lock_guard<std::mutex> lk_guard (m_mutex);
 
-      T_PAYLOAD &ent = m_response_payloads[a_rsn];
-      ent = std::move (a_payload);
+      payload_or_error_type &ent = m_response_payloads[a_rsn];
+      ent.m_payload = std::move (a_payload);
+      ent.m_error = m_no_error;
+    }
+    m_condvar.notify_all (); // all because there is more than one thread waiting for data
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_error (response_sequence_number a_rsn, T_ERROR &&a_error)
+  {
+    {
+      std::lock_guard<std::mutex> lockg (m_mutex);
+
+      payload_or_error_type &ent = m_response_payloads[a_rsn];
+      ent.m_error = std::move (a_error);
     }
-    m_condvar.notify_all ();
+    m_condvar.notify_all (); // all because there is more than one thread waiting for data
   }
 
-  template <typename T_PAYLOAD>
-  T_PAYLOAD
-  response_broker<T_PAYLOAD>::get_response (response_sequence_number a_rsn)
+  template <typename T_PAYLOAD, typename T_ERROR>
+  std::tuple<T_PAYLOAD, T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::get_response (response_sequence_number a_rsn)
   {
     return get_bucket (a_rsn).get_response (a_rsn);
   }
 
-  template <typename T_PAYLOAD>
-  T_PAYLOAD
-  response_broker<T_PAYLOAD>::bucket::get_response (response_sequence_number a_rsn)
+  template <typename T_PAYLOAD, typename T_ERROR>
+  std::tuple<T_PAYLOAD, T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::get_response (response_sequence_number a_rsn)
   {
     std::unique_lock<std::mutex> ulock (m_mutex);
-    T_PAYLOAD payload;
+    std::tuple<T_PAYLOAD, T_ERROR> payload_or_error;
 
-    auto condvar_pred = [this, &payload, a_rsn] ()
+    auto condvar_pred = [this, &payload_or_error, a_rsn] ()
```

Comment:
Please, be sure that the thread does not hang - use a flag and cond_wait with timeout for instance.

Reply:
Indeed `get_response` might hang when the connection is dropped for whatever reason.
Added a new flag - `m_terminate` - the `bucket` and set it for all buckets in the `request_sync_client_server` dtor after the sending/receiving of network messages is terminated.